File: LeetCode/Problems/Python/39. Combination Sum.py
#backtracking 
#Time: O(n^k)
#Space: O(k), when worst case, [1,1,1,1] 4, it will run 4 times recursive
class Solution:
    def dfs(self, candidates, target, result, tempSet, start):
        if target < 0:
            return
        if target == 0:
            result.append(tempSet)
            return
        for i in range(start, len(candidates)):
            self.dfs(candidates, target-candidates[i], result, tempSet+[candidates[i]], i)

        
        
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        if not candidates:
            return []
        
        candidates.sort()
        result = []
        self.dfs(candidates, target, result, [], 0)
        return result

File: LeetCode/Problems/C++/39. Combination Sum.cpp
//Time complexity is O(N^target) where N is a length of candidates array.
//Space complexity is O(target). if the smallest candidate is 1, the recursion will go up to target depth

class Solution {
public:
    void searchComb(vector<int>& candid, int idx, int target, vector<int>& curr, vector<vector<int>> &result) {
        if(target == 0){
            result.push_back(curr);
        }
        if(target < 0) return;
        
        for(int i=idx;i<candid.size();i++){
            curr.push_back(candid[i]);
            searchComb(candid, i, target-candid[i], curr, result);
            curr.pop_back();
        }   
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(!candidates.size()) return {};
        
        vector<vector<int>> result;
        vector<int> curr;
        searchComb(candidates, 0, target, curr, result);
        return result;
        
    }
};